compiler/rustc_codegen_ssa/src/back/linker.rs

```diff
@@ -860,6 +876,8 @@ impl<'a> Linker for GccLinker<'a> {
             self.link_arg("-M").link_arg(path);
         } else if is_windows {
             self.link_arg(path);
+        } else if crate_type == CrateType::Executable {
+            self.link_arg("--dynamic-list").link_arg(path);
         } else {
             let mut arg = OsString::from("--version-script=");
             arg.push(path);
```

Comment:
Maybe we should merge the if else chains for writing the temporary file and for passing the arguments. They both use the exact same conditions.

Reply:
We can have something like

```rust
if is_like_windows {
    // write def file
    // link arg
} else if is_like_darwin {
    // write linker script
    // link arg
} else if executable && !is_like_solaris {
    // write symbol list
    // link arg
} else {
    // write linker script
    if is_like_solaris {
        // link arg
    } else {
        // link arg
    }
}
```